packages/client-assertion-validation/src/validation.ts

```diff
@@ -0,0 +1,259 @@
+      decoded.payload.jti
+    );
+    const { errors: iatErrors, data: validatedIat } = validateIat(
+      decoded.payload.iat
+    );
+    const { errors: expErrors, data: validatedExp } = validateExp(
+      decoded.payload.exp
+    );
+    const { errors: issErrors, data: validatedIss } = validateIss(
+      decoded.payload.iss
+    );
+    const { errors: subErrors, data: validatedSub } = validateSub(
+      decoded.payload.sub,
+      clientId
+    );
+    const { errors: purposeIdErrors, data: validatedPurposeId } =
+      validatePurposeId(decoded.payload.purposeId);
+    const { errors: kidErrors, data: validatedKid } = validateKid(
+      decoded.header.kid
+    );
+    const { errors: audErrors, data: validatedAud } = validateAudience(
+      decoded.payload.aud
+    );
+    const { errors: algErrors, data: validatedAlg } = validateAlgorithm(
+      decoded.header.alg
+    );
+    const { errors: digestErrors, data: validatedDigest } = validateDigest(
+      decoded.payload.digest
+    );
+    if (
+      !jtiErrors &&
+      !iatErrors &&
+      !expErrors &&
+      !issErrors &&
+      !subErrors &&
+      !purposeIdErrors &&
+      !kidErrors &&
+      !audErrors &&
+      !algErrors &&
+      !digestErrors
+    ) {
+      const result: ClientAssertion = {
+        header: {
+          kid: validatedKid,
+          alg: validatedAlg,
+        },
+        payload: {
+          sub: validatedSub,
+          purposeId: validatedPurposeId,
+          jti: validatedJti,
+          iat: validatedIat,
+          iss: validatedIss,
+          aud: validatedAud,
+          exp: validatedExp,
+          digest: validatedDigest,
+        },
+      };
+      return successfulValidation(result);
+    }
+    return failedValidation([
+      jtiErrors,
+      iatErrors,
+      expErrors,
+      issErrors,
+      subErrors,
+      purposeIdErrors,
+      kidErrors,
+      audErrors,
+      algErrors,
+      digestErrors,
+    ]);
+  } catch (error) {
+    return failedValidation([unexpectedClientAssertionPayload()]);
+  }
+};
+
+export const verifyClientAssertionSignature = (
+  clientAssertionJws: string,
+  key: Key
+): ValidationResult<JwtPayload> => {
+  try {
+    const result = verify(clientAssertionJws, key.publicKey, {
+      algorithms: [key.algorithm],
+    });
+
+    // TODO: no idea when result is a string
+    if (typeof result === "string") {
+      return failedValidation([
+        invalidClientAssertionSignatureType(typeof result),
+      ]);
+    }
+    return successfulValidation(result);
+  } catch (error: unknown) {
+    if (error instanceof TokenExpiredError) {
+      return failedValidation([tokenExpiredError()]);
+    } else if (error instanceof NotBeforeError) {
+      return failedValidation([notBeforeError()]);
+    } else if (error instanceof JsonWebTokenError) {
+      // TODO: this might overlap with invalidClientAssertionFormat raised inside verifyClientAssertion
```

Comment:
The doubt was about not reaching that case at all, because in that situation `invalidClientAssertionFormat` is raised in the `verifyClientAssertion` function, which we will call before.

Reply:
Update: we keep different errors to distinguish where they are raised.